src/hotspot/share/gc/parallel/psParallelCompact.cpp

```diff
@@ -1432,68 +1432,36 @@ void PSParallelCompact::summarize_spaces_quick()
-  const RegionData* region = _summary_data.addr_to_region_ptr(dense_prefix_end);
-  const idx_t dense_prefix_bit = _mark_bitmap.addr_to_bit(dense_prefix_end);
-  if (dead_space_crosses_boundary(region, dense_prefix_bit)) {
-    // Only enough dead space is filled so that any remaining dead space to the
-    // left is larger than the minimum filler object.  (The remainder is filled
-    // during the copy/update phase.)
-    //
-    // The size of the dead space to the right of the boundary is not a
-    // concern, since compaction will be able to use whatever space is
-    // available.
-    //
-    // Here '||' is the boundary, 'x' represents a don't care bit and a box
-    // surrounds the space to be filled with an object.
-    //
-    // In the 32-bit VM, each bit represents two 32-bit words:
-    //                              +---+
-    // a) beg_bits:  ...  x   x   x | 0 | ||   0   x  x  ...
-    //    end_bits:  ...  x   x   x | 0 | ||   0   x  x  ...
-    //                              +---+
-    //
-    // In the 64-bit VM, each bit represents one 64-bit word:
-    //                              +------------+
-    // b) beg_bits:  ...  x   x   x | 0   ||   0 | x  x  ...
-    //    end_bits:  ...  x   x   1 | 0   ||   0 | x  x  ...
-    //                              +------------+
-    //                          +-------+
-    // c) beg_bits:  ...  x   x | 0   0 | ||   0   x  x  ...
-    //    end_bits:  ...  x   1 | 0   0 | ||   0   x  x  ...
-    //                          +-------+
-    //                      +-----------+
-    // d) beg_bits:  ...  x | 0   0   0 | ||   0   x  x  ...
-    //    end_bits:  ...  1 | 0   0   0 | ||   0   x  x  ...
-    //                      +-----------+
-    //                          +-------+
-    // e) beg_bits:  ...  0   0 | 0   0 | ||   0   x  x  ...
-    //    end_bits:  ...  0   0 | 0   0 | ||   0   x  x  ...
-    //                          +-------+
-
-    // Initially assume case a, c or e will apply.
-    size_t obj_len = CollectedHeap::min_fill_size();
-    HeapWord* obj_beg = dense_prefix_end - obj_len;
-
-#ifdef  _LP64
-    if (MinObjAlignment > 1) { // object alignment > heap word size
-      // Cases a, c or e.
-    } else if (_mark_bitmap.is_obj_end(dense_prefix_bit - 2)) {
-      // Case b above.
-      obj_beg = dense_prefix_end - 1;
-    } else if (!_mark_bitmap.is_obj_end(dense_prefix_bit - 3) &&
-               _mark_bitmap.is_obj_end(dense_prefix_bit - 4)) {
-      // Case d above.
-      obj_beg = dense_prefix_end - 3;
-      obj_len = 3;
-    }
-#endif  // #ifdef _LP64
+  RegionData* const region = _summary_data.addr_to_region_ptr(dense_prefix_end);
+  idx_t const dense_prefix_bit = _mark_bitmap.addr_to_bit(dense_prefix_end);
 
+  if (region->partial_obj_size() != 0 ||
+      _mark_bitmap.is_obj_beg(dense_prefix_bit)) {
+    // next region starts with live bytes
+    return;
+  }
+
+  if (_mark_bitmap.is_obj_end(dense_prefix_bit - 2)) {
```

Comment:
I would almost recommend inverting this condition with a return, but feel free to ignore.